File: sdks/typescript/src/time_values.ts
function asDate(value: Date | string): Date {
  const date = value instanceof Date ? new Date(value.getTime()) : new Date(value);
  if (Number.isNaN(date.getTime())) throw new TypeError("invalid RFC3339 timestamp");
  return date;
}

function localRfc3339(date: Date): string {
  const offsetMinutes = -date.getTimezoneOffset();
  const sign = offsetMinutes >= 0 ? "+" : "-";
  const absolute = Math.abs(offsetMinutes);
  const hours = String(Math.floor(absolute / 60)).padStart(2, "0");
  const minutes = String(absolute % 60).padStart(2, "0");
  const shifted = new Date(date.getTime() + offsetMinutes * 60_000);
  return `${shifted.toISOString().replace("Z", "")}${sign}${hours}:${minutes}`;
}

/** Current Unix timestamp in seconds, returning zero only on invalid Date state. */
export function nowSecs(): number {
  return Math.floor(Date.now() / 1_000);
}

/** Current Unix timestamp in milliseconds. */
export function nowMillis(): number {
  return Date.now();
}

/** Current system-local timestamp with an explicit RFC3339 offset. */
export function nowLocal(): string {
  return localRfc3339(new Date());
}

/** Convert an instant to the system-local RFC3339 representation. */
export function toLocal(value: Date | string): string {
  return localRfc3339(asDate(value));
}

/** Serialize a UTC/offset instant using the system-local RFC3339 offset. */
export function localRfc3339Serialize(value: Date | string): string {
  return toLocal(value);
}

/** Parse a local or UTC RFC3339 value and return the same instant as Date. */
export function localRfc3339Deserialize(value: string): Date {
  return asDate(value);
}

export function optionLocalRfc3339Serialize(value: Date | string | null | undefined): string | null {
  return value == null ? null : toLocal(value);
}

export function optionLocalRfc3339Deserialize(value: string | null | undefined): Date | null {
  return value == null ? null : localRfc3339Deserialize(value);
}

File: sdks/python/src/echo_agent_sdk/time_values.py
from __future__ import annotations

import time
from datetime import datetime, timezone


def now_secs() -> int:
    return time.time_ns() // 1_000_000_000


def now_millis() -> int:
    return time.time_ns() // 1_000_000


def now_local() -> datetime:
    return datetime.now().astimezone()


def to_local(value: datetime) -> datetime:
    if value.tzinfo is None:
        raise ValueError("timestamp must be timezone-aware")
    return value.astimezone()


def local_rfc3339_serialize(value: datetime) -> str:
    return to_local(value).isoformat()


def local_rfc3339_deserialize(value: str) -> datetime:
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as error:
        raise ValueError("invalid RFC3339 timestamp") from error
    if parsed.tzinfo is None:
        raise ValueError("timestamp must include an offset")
    return parsed.astimezone(timezone.utc)


def option_local_rfc3339_serialize(value: datetime | None) -> str | None:
    return None if value is None else local_rfc3339_serialize(value)


def option_local_rfc3339_deserialize(value: str | None) -> datetime | None:
    return None if value is None else local_rfc3339_deserialize(value)
